apm-agent-plugins/apm-log-shader-plugin/apm-log-shader-plugin-common/src/main/java/co/elastic/apm/agent/log/shader/Utils.java

```diff
@@ -0,0 +1,80 @@
+/*-
+ * #%L
+ * Elastic APM Java agent
+ * %%
+ * Copyright (C) 2018 - 2020 Elastic and contributors
+ * %%
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.log.shader;
+
+import co.elastic.apm.agent.impl.GlobalTracer;
+import co.elastic.apm.agent.logging.LoggingConfiguration;
+
+import javax.annotation.Nullable;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+public class Utils {
+
+    private static final String SHADE_FILE_EXTENSION = ".ecs.json";
+    private static final LoggingConfiguration config = GlobalTracer.requireTracerImpl().getConfig(LoggingConfiguration.class);
```

Comment:
As this class has a dependency (`LoggingConfiguration`), I'd prefer it not to be a static utility class but to have a constructor argument for `LoggingConfiguration`.
It'd simplify `UtilsTest` (or at least make it more lightweight) in that it doesn't need to extend from `AbstractInstrumentationTest` and you could just inject a mock instance for  `LoggingConfiguration`.

Also, consider creating a constructor argument for `LoggingConfiguration` in `AbstractLogShadingHelper`. Instead of making the implementations singletons, have a look at `co.elastic.apm.agent.jdbc.JdbcInstrumentation#getJdbcHelper`.

By using that pattern, the fact that advices are static does not leak into the rest of the codebase which can use regular constructor-based "dependency-injection". I think it makes sense to follow a similar pattern throughout the codebase so that it's more cohesive and can serve as a reference for new plugins.

Reply:
Actually, I think `co.elastic.apm.agent.log4j2.Log4j2LogShadingHelper#instance` can probably stay as-is and just change `private static final Log4j2LogShadingHelper INSTANCE = new Log4j2LogShadingHelper();` to `private static final Log4j2LogShadingHelper INSTANCE = new Log4j2LogShadingHelper(GlobalTracer.requireTracerImpl());`

I like the fact that it's a final field rather than a volatile field in `JdbcInstrumentation`

Reply:
`Utils` is a static utility by its nature. The argument of having a "dependency" shouldn't interfere with that. Depending on the logging framework as in `LoggerFactory.getLogger()`, for example, would be the same but would not exclude a class from being a static utility. 
Doing what you suggest can create the exact opposite problem of enforcing the utility user to introduce this dependency, which it might don't even need for anything else. And given that it is a utility, this will affect multiple users.

Reply:
IMHO, a static utility should not depend on another class that is stateful. But what do you think about just adding a `String shadeFileDestinationDir` parameter that the caller can read from `config.getLogEcsFormattingDestinationDir()`?

Reply:
There is such method already: https://github.com/elastic/apm-agent-java/pull/1261/files#diff-2f5c514f954a475ff220cebc3a8551dce7915534f14d15bcffa86021ad108403R322
but it only returns the configured value or `null`.

The documentation for the `log_ecs_reformatting_dir` config option says: 
```
If `log_ecs_reformatting` is set to `SHADE` or `REPLACE`, 
the shade log files will be written alongside 
the original logs in the same directory by default. 
Use this configuration in order to write the shade logs 
into an alternative destination. Omitting this config or 
setting it to an empty string will restore the default 
behavior. If relative path is used, this path will be used 
relative to the original logs directory.
```
In order to support that, there is some logic required, as well as an input argument for the original log file, so I felt this is not very appropriate to be located within the config class.